### DISCUSSION TITLE
How to build a dataset navigator

### DISCUSSION THREAD

**Discussion by davebound** [rrousselGit/riverpod#1413] [Q&A]
Hi,

I'm stuck on figuring out the best way to achieve a simple thing.

Part of my app needs to allow the user to step through records in a database using Next/Previous buttons

![image](https://user-images.githubusercontent.com/18735591/168571963-b3d7f936-05e7-49e9-bdf2-ff30adcf735e.png)

I thought one way to do this would be to have a FutureProvider for the Record that makes async calls to web api.

The query it sends to the API depends on which button was pressed (Next or Previous), so I thought I could make it watch another provider that encodes the command:

```
enum Cmd {
  none,
  next,
  previous
}
gProvCmdProvider = StateProvider<Cmd>((ref)=>Cmd.none)


gProvRecordProvider = FutureProvider.autodispose<Record>((ref) async {

  final cmd = ref.watch(gProvCmdProvider);

  // web api query depends on the command
  final qry = buildQry(cmd);
  final data = await fetch(qry);

  return Record(data);
})
```

onPressed handlers in the UI could then do

`ref.watch(gProvCmd).state = Cmd.next`

or

`ref.watch(gProvCmd).state = Cmd.prev`

Questions:
- Once gProvRecordProvider  has consumed the command, I want to set gProvCmdProvider back to Cmd.none. But I cant do that in gProvRecordProvider because circular reference type issue. I would like gProvCmdProvider to be a "one-shot" that auto-resets itself to "none"
- Pressing next or prev buttons might cause the server to return "no more records" error, in which case I want gProvRecordProvider to return the same Record it returned last time. Does Riverpod store that info anywhere?
- I also want to implement a Goto button which prompts the user for a record Id to jump to. I could encode that either in another provider like gProvCmdProvider, or in its own provider - but once again, it needs the ability to cope with resetting itself or handling the case where record not found

Many thanks

**Comment by rrousselGit** [comment]
Hello!
That's not the correct architecture. You probably don't want that `Cmd` enum. Instead use an `int` for the page index

```dart
final page = StateProvider<int>((ref) => 0);
```

which you increment/decrement on prev/next tap

**Reply by davebound** [reply]
Hi Remi,

They're not pages. They're record ids, and they won't be contiguous. So, next and prev is definitely what I need.

I guess in general is what I'm looking for is an architecture that can be extended to other "commands" e.g. scroll 10 ahead

Thx

**Reply by rrousselGit** [reply]
The principle is the same. That Cmd is never going to be what you want

You should have something equivalent to the page index. Such as infos from the previous fetch

**Reply by davebound** [reply]
Thx, but I'm still confused and its not really answering the questions from my orig post.

It makes sense that my gProvRecordProvider  is a FutureProvider right? Because it needs to do an async call to server.

Its also true that I need to inject data  into gProvRecordProvider  so that it knows what data to request from the server. So doesn't it make sense that that injection is done by making gProvRecordProvider  dependent on another provider that is controlled by the buttons? Whether that provides an enum (Cmd), int (pageId) is, I think, incidental to my questions.

Or are you saying its fundamentally the wrong approach to inject data using a riverpod dependency? It feels no different to me that the Todo example, where you inject the full todo list into the completed todo list. So what am I missing?

How would you do what I'm trying to do? Seems like a very common use case, no?

**Reply by rrousselGit** [reply]
I'm certain about it. The Cmd enum you're currently using is truly the source of your problem. As you explained here:

> Once gProvRecordProvider has consumed the command, I want to set gProvCmdProvider back to Cmd.none. But I cant do that in gProvRecordProvider because circular reference type issue. I would like gProvCmdProvider to be a "one-shot" that auto-resets itself to "none"

In Riverpod you don't "consume an event". If you were to use something like a page index (or whatever equivalent there is in your system), this problem wouldn't exist.

A page index/record ID/... is not an event. It's a value, and that values doesn't have to "reset to none". That value could even be serialized and stored in a local DB
